tools/protodoc/protodoc.py

```diff
@@ -184,71 +243,153 @@ def StripLeadingSpace(s):
+  for field in msg.field:
+    if field.HasField('oneof_index'):
+      type_context.oneof_fields[field.oneof_index].append(field.name)
+  return '\n'.join(
+      FormatFieldAsDefinitionListItem(
+          type_context.Extend([2, index], field.name), field)
       for index, field in enumerate(msg.field)) + '\n'
 
 
-def FormatMessage(source_code_info, path, msg):
-  """Format a MessageDescriptorProto as RST section.
+def FormatMessage(type_context, msg):
+  """Format a DescriptorProto as RST section.
 
   Args:
-    source_code_info: SourceCodeInfo object.
-    path: a list of path indexes as per
-      https://github.com/google/protobuf/blob/a08b03d4c00a5793b88b494f672513f6ad46a681/src/google/protobuf/descriptor.proto#L717.
-    msg: MessageDescriptorProto.
+    type_context: contextual information for message/enum/field.
+    msg: DescriptorProto.
   Returns:
     RST formatted section.
   """
-  anchor = FormatAnchor(MessageCrossRefLabel(msg.name))
-  header = FormatHeader('-', msg.name)
-  comment = source_code_info.LeadingCommentPathLookup(path)
+  # Skip messages synthesized to represent map types.
+  if msg.options.map_entry:
+    return ''
+  # We need to do some extra work to recover the map type annotation from the
+  # synthesized messages.
+  type_context.map_typenames = {
+      '%s.%s' % (type_context.name, nested_msg.name): 'map<%s, %s>' % tuple(
+          map(
+              functools.partial(FormatFieldType, type_context),
+              nested_msg.field))
+      for nested_msg in msg.nested_type if nested_msg.options.map_entry
+  }
+  nested_msgs = '\n'.join(
+      FormatMessage(
+          type_context.Extend([3, index], nested_msg.name), nested_msg)
+      for index, nested_msg in enumerate(msg.nested_type))
+  nested_enums = '\n'.join(
+      FormatEnum(
+          type_context.Extend([4, index], nested_enum.name), nested_enum)
+      for index, nested_enum in enumerate(msg.enum_type))
+  anchor = FormatAnchor(MessageCrossRefLabel(type_context.name))
+  header = FormatHeader('-', type_context.name)
+  comment = type_context.LeadingCommentPathLookup()
   return anchor + header + comment + FormatMessageAsJson(
-      msg) + FormatMessageAsDefinitionList(source_code_info, path, msg)
+      type_context, msg) + FormatMessageAsDefinitionList(
+          type_context, msg) + nested_msgs + '\n' + nested_enums
+
+
+def FormatEnumValueAsDefinitionListItem(type_context, enum_value):
+  """Format a EnumValueDescriptorProto as RST definition list item.
+
+  Args:
+    type_context: contextual information for message/enum/field.
+    enum_value: EnumValueDescriptorProto.
+  Returns:
+    RST formatted definition list item.
+  """
+  anchor = FormatAnchor(EnumValueCrossRefLabel(type_context.name))
+  default_comment = '*(DEFAULT)* ' if enum_value.number == 0 else ''
+  comment = default_comment + UNICODE_INVISIBLE_SEPARATOR + type_context.LeadingCommentPathLookup(
+  )
+  return anchor + '%s (%d)' % (enum_value.name,
```

Comment:
I would omit number. Unless we want to explicitly say this is proto and so enums matter.